pkg/discovery/worker/compliance/affinity_processor.go

```diff
@@ -61,7 +65,7 @@ func (am *AffinityProcessor) processAffinityPerPod(pod *api.Pod, podsNodesMap ma
 
 	hostNode := podsNodesMap[pod]
 	for _, node := range am.nodes {
-		if matchesNodeSelector(pod, node) && matchesNodeAffinity(pod, node) {
+		if matchesNodeAffinity(pod, node) && matchesPvNodeAffinity(pvNodeSelectorTerms, node) {
```

Comment:
Because this condition actually introduces a bug we may never test out before. 
Let's say the cluster has 2 worker nodes `worker0` and `worker1`, `worker0` has the label `disktype=ssd` and `xxx=yyy`, and `worker1` has none of them. 
A pod(without PV) with pod nodeAffinity as `disktype=ssd` and nodeSelector as `xxx=yyy` comes in , it will be running on `worker0` for sure. Then we remove the label `xxx=yyy` from `worker0` and add it to `worker1`, we expect there is  no move action generated as `worker1` doesn't meet the pod's nodeAffinity. But in fact, we get a move action, that's just because we have this `matchesNodeSelector` condition. When it checks the `worker0`, `worker0` doesn't have the label `xxx=yyy` now, it ends up returning false.  when it checks `worker1`, `worker1` doesn't meet the nodeAffinity,it also return false.
Finally, the pod itself, `worker0` and `worker1` all of them skip adding VMPM commodity. On the market side, it will only consider `worker2` has the label `xxx=yyy` that meets the pod's nodeSelector, so it generates a move action. 

I don't think we need that condition when we check the `nodeAffnity`, the handling of `nodeSeletor` is already done with the `label` commodity.

Reply:
If we remove this check we will remove handling of specific cases, for example the case where a pod does not specify a nodeSelector, but specifies a nodeAffinity? 
The scenario you mentioned might be a bug, and ideally the result should be an action indicating `misconfiguration`.

Reply:
@irfanurrehman , could you please elaborate a little bit about the specific case you mentioned?  I'd love to test it out based on my code change. From my understanding, we added this `matchesNodeSelector` expression when we didn't have the `label` commodity.  

Based on my code change, the scenario I mentioned above is getting a `reconfigure` action, only the description of that action doesn't seem right.  

![image](https://user-images.githubusercontent.com/61252360/187446602-136c9703-0eda-4169-b98b-a8b9e1ce9b11.png)
The pod yaml for your reference:
```
[root@IBM-PF3888SM pvaffinity]# k get pods nginx-deployment-86f6856fbd-fjbd8 -o yaml
apiVersion: v1
kind: Pod
metadata:
  annotations:
    k8s.v1.cni.cncf.io/network-status: |-
      [{
          "name": "openshift-sdn",
          "interface": "eth0",
          "ips": [
              "10.254.19.138"
          ],
          "default": true,
          "dns": {}
      }]
    k8s.v1.cni.cncf.io/networks-status: |-
      [{
          "name": "openshift-sdn",
          "interface": "eth0",
          "ips": [
              "10.254.19.138"
          ],
          "default": true,
          "dns": {}
      }]
    openshift.io/scc: restricted
  creationTimestamp: "2022-08-30T13:12:23Z"
  generateName: nginx-deployment-86f6856fbd-
  labels:
    app: nginx
    pod-template-hash: 86f6856fbd
  name: nginx-deployment-86f6856fbd-fjbd8
  namespace: default
  ownerReferences:
  - apiVersion: apps/v1
    blockOwnerDeletion: true
    controller: true
    kind: ReplicaSet
    name: nginx-deployment-86f6856fbd
    uid: 5c5a95de-e464-4892-8326-52b20a078bf0
  resourceVersion: "537094"
  uid: 59f3c75d-b1ba-467a-838f-b333d278f0ce
spec:
  affinity:   <------------pod has nodeAffinity
    nodeAffinity:
      requiredDuringSchedulingIgnoredDuringExecution:
        nodeSelectorTerms:
        - matchExpressions:
          - key: disktype
            operator: In
            values:
            - ssd
  containers:
  - env:
    - name: RUN_TYPE
      value: cpu
    - name: CPU_PERCENT
      value: "100"
    image: beekman9527/cpumemload:latest
    imagePullPolicy: Always
    name: nginx
    resources:
      limits:
        cpu: 1510m
      requests:
        cpu: 10m
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /mnt
      name: pxvol
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: kube-api-access-h8v4f
      readOnly: true
  dnsPolicy: ClusterFirst
  enableServiceLinks: true
  imagePullSecrets:
  - name: turboregistrykey
  - name: turbocred
  nodeName: worker0.brutes.cp.fyre.ibm.com
  nodeSelector:
    xxx: yyy
  preemptionPolicy: PreemptLowerPriority
  priority: 0
  restartPolicy: Always
  schedulerName: default-scheduler
  securityContext: {}
  serviceAccount: default
  serviceAccountName: default
  terminationGracePeriodSeconds: 30
  tolerations:
  - effect: NoExecute
    key: node.kubernetes.io/not-ready
    operator: Exists
    tolerationSeconds: 300
  - effect: NoExecute
    key: node.kubernetes.io/unreachable
    operator: Exists
    tolerationSeconds: 300
  - effect: NoSchedule
    key: node.kubernetes.io/memory-pressure
    operator: Exists
  volumes:
  - name: pxvol
    persistentVolumeClaim:
      claimName: task-pv-claim
  - name: kube-api-access-h8v4f
    projected:
      defaultMode: 420
      sources:
      - serviceAccountToken:
          expirationSeconds: 3607
          path: token
      - configMap:
          items:
          - key: ca.crt
            path: ca.crt
          name: kube-root-ca.crt
      - downwardAPI:
          items:
          - fieldRef:
              apiVersion: v1
              fieldPath: metadata.namespace
            path: namespace
      - configMap:
          items:
          - key: service-ca.crt
            path: service-ca.crt
          name: openshift-service-ca.crt
```
```
[root@IBM-PF3888SM pvaffinity]# k get pvc -o yaml
apiVersion: v1
items:
- apiVersion: v1
  kind: PersistentVolumeClaim
  metadata:
    annotations:
      pv.kubernetes.io/bind-completed: "yes"
      pv.kubernetes.io/bound-by-controller: "yes"
    creationTimestamp: "2022-08-30T12:53:41Z"
    finalizers:
    - kubernetes.io/pvc-protection
    name: task-pv-claim
    namespace: default
    resourceVersion: "530232"
    uid: 67578f59-5a78-4005-9639-4843da98bdf5
  spec:
    accessModes:
    - ReadWriteOnce
    resources:
      requests:
        storage: 1Gi
    storageClassName: manual
    volumeMode: Filesystem
    volumeName: task-pv-volume
  status:
    accessModes:
    - ReadWriteOnce
    capacity:
      storage: 1Gi
    phase: Bound
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""
```
```
[root@IBM-PF3888SM pvaffinity]# k get pv task-pv-volume -o yaml
apiVersion: v1
kind: PersistentVolume
metadata:
  annotations:
    pv.kubernetes.io/bound-by-controller: "yes"
  creationTimestamp: "2022-08-30T12:53:10Z"
  finalizers:
  - kubernetes.io/pv-protection
  labels:
    type: local
  name: task-pv-volume
  resourceVersion: "530230"
  uid: aed1a925-d5c5-4e67-9dbe-fa12c8ad6c4a
spec:
  accessModes:
  - ReadWriteOnce
  capacity:
    storage: 1Gi
  claimRef:
    apiVersion: v1
    kind: PersistentVolumeClaim
    name: task-pv-claim
    namespace: default
    resourceVersion: "530228"
    uid: 67578f59-5a78-4005-9639-4843da98bdf5
  hostPath:   <-------- PV doesn't have nodeAffninty
    path: /tmp
    type: ""
  persistentVolumeReclaimPolicy: Retain
  storageClassName: manual
  volumeMode: Filesystem
status:
  phase: Bound
```

The pod is running on `worker0` when it started, now I move the `xxx=yyy` from `worker0` to `worker1`, in the current status, there is no node that can match all of the pods' requirements.  there is no `move` action and only a `reconfigure` action is generated as expected. 
```
[root@IBM-PF3888SM pvaffinity]# k get node --show-labels worker0.brutes.cp.fyre.ibm.com  worker1.brutes.cp.fyre.ibm.com  worker2.brutes.cp.fyre.ibm.com 
NAME                             STATUS   ROLES    AGE   VERSION            LABELS
worker0.brutes.cp.fyre.ibm.com   Ready    worker   23h   v1.21.11+6b3cbdd   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/os=linux,disktype=ssd,kubernetes.io/arch=amd64,kubernetes.io/hostname=worker0.brutes.cp.fyre.ibm.com,kubernetes.io/os=linux,node-role.kubernetes.io/worker=,node.openshift.io/os_id=rhcos,topology.kubernetes.io/zone=us-east-1
worker1.brutes.cp.fyre.ibm.com   Ready    worker   23h   v1.21.11+6b3cbdd   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/os=linux,kubernetes.io/arch=amd64,kubernetes.io/hostname=worker1.brutes.cp.fyre.ibm.com,kubernetes.io/os=linux,node-role.kubernetes.io/worker=,node.openshift.io/os_id=rhcos,topology.kubernetes.io/zone=us-east-1,xxx=yyy
worker2.brutes.cp.fyre.ibm.com   Ready    worker   23h   v1.21.11+6b3cbdd   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/os=linux,kubernetes.io/arch=amd64,kubernetes.io/hostname=worker2.brutes.cp.fyre.ibm.com,kubernetes.io/os=linux,node-role.kubernetes.io/worker=,node.openshift.io/os_id=rhcos,topology.kubernetes.io/zone=us-west-1
```

Reply:
Ack, I was missing the handling of the nodeSelectors by the label commodities. Thanks for explaining @kevinwangcn

Reply:
@irfanurrehman  thanks for reviewing the code and approving the PR.